File: javascript/DataStructures/Stack.js
class Node {
    constructor(value) {
        this.data = value;
        this.next = null;
    }
}

class Stack {
    constructor() {
        this.head = null;
    }

    push(value) {
        let newNode = new Node(value);
        if (this.head === null) {
            this.head = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
    }

    pop() {
        if (this.isEmpty()) {
            throw new Error("Empty stack. Cannot perform pop!")
        }
        if (this.head.next === null) {
            this.head = null;
        } else {
            this.head = this.head.next;
        }
    }

    peek() {
        return this.head.data
    }

    isEmpty() {
        return this.head === null;
    }
}

// MAIN //
const s = new Stack();
s.push(1);
s.push(2);
s.push(3);
s.push(4);

while (!s.isEmpty()) {
    console.log(s.peek());
    s.pop();
}

File: go/DataStructures/Stack/Stack.go
package Stack

import (
	"errors"
)

type Node struct {
	data int
	next *Node
}

func NewNode(value int) *Node {
	return &Node{
		data: value,
		next: nil,
	}
}

type Stack struct {
	head *Node
}

func NewStack() *Stack {
	return &Stack{
		head: nil,
	}
}

func (s *Stack) Push(value int) {
	newNode := NewNode(value)

	if s.IsEmpty() {
		s.head = newNode
	} else {
		newNode.next = s.head
		s.head = newNode
	}
}

func (s *Stack) Pop() error {
	if s.IsEmpty() {
		return errors.New("Empty stack. Can not perform Pop")
	}
	s.head = s.head.next

	return nil
}

func (s *Stack) Peek() int {
    if s.IsEmpty() {
        return -1
    }

    return s.head.data
}

func (s *Stack) IsEmpty() bool {
	return s.head == nil
}
